File: console/src/main/java/com/damon140/ur/PlayArea.java
/*
 * Copyright 2023 Damon van Opdorp
 *
 * Licensed under GNU General Public License v3.0.
 *
 * https://www.gnu.org/licenses/gpl-3.0.en.html
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.damon140.ur;

import java.security.NoSuchAlgorithmException;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.stream.Collectors;

public class PlayArea {

    public static final int COUNTERS_PER_PLAYER = 7;

    private final Map<Square, Team> counters;

    private final Map<Team, Integer> completedCounters;

    private Team currentTeam = Team.white; // white starts

    public PlayArea() throws NoSuchAlgorithmException {
        counters = new HashMap<>();
        completedCounters = new TreeMap<>();
        completedCounters.put(Team.black, 0);
        completedCounters.put(Team.white, 0);
    }

    public int completedCount(Team team) {
        return completedCounters.get(team);
    }

    public Set<Square> countersForTeam(Team team) {
        return this.counters.entrySet()
                .stream()
                .filter(entry -> team == entry.getValue())
                .map(Map.Entry::getKey)
                .collect(Collectors.toSet());
    }

    public Team get(Square square) {
        return this.counters.get(square);
    }

    public boolean occupied(Square square) {
        return this.counters.containsKey(square);
    }

    public Team currentTeam() {
        return this.currentTeam;
    }

    public void swapTeam() {
        this.currentTeam = this.currentTeam.other();
    }

    public boolean allCountersStarted(Team team) {
        return COUNTERS_PER_PLAYER == completedCounters.get(team);
    }

    public int unstartedCount(Team team) {
        long inProgressCounters = this.counters.entrySet()
                .stream()
                .filter(e -> team == e.getValue())
                .count();
        int finishedCounters = this.completedCounters.get(team);

        return COUNTERS_PER_PLAYER - (finishedCounters + (int) inProgressCounters);
    }

    public boolean allStartedOrComplete(Team team) {
        return 0 == this.unstartedCount(team);
    }

    public int inPlayCount(Team team) {
        return (int) counters.values().stream().filter(t -> team == t).count();
    }

    /** All counters currently on the board from both teams. */
    public int inPlayCount() {
        return this.counters.size();
    }

    public void move(Square fromSquare, Square newSquare, Team team) {
        counters.remove(fromSquare);
        if (newSquare != Square.off_board_finished) {
            counters.put(newSquare, team);
        } else {
            completedCounters.put(team, 1 + completedCounters.get(team));
        }
    }

    public boolean allCompleted(Team team) {
        return PlayArea.COUNTERS_PER_PLAYER == this.completedCounters.get(team);
    }

}

File: web/src/main/kotlin/ur/PlayArea.kt
/*
 * Copyright 2023 Damon van Opdorp
 *
 * Licensed under GNU General Public License v3.0.
 *
 * https://www.gnu.org/licenses/gpl-3.0.en.html
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.damon140.ur

class PlayArea(private var currentTeam: Team) {
    private val counters: MutableMap<Square, Team>
    private val completedCounters: MutableMap<Team, Int>

    init {
        counters = HashMap()
        completedCounters = HashMap<Team, Int>()
        completedCounters[Team.Black] = 0
        completedCounters[Team.White] = 0
    }

    fun moveTakes(team: Team, fromSquare: Square, toSquare: Square): Boolean {
        return teamHasCounterOn(team, fromSquare)
                && teamHasCounterOn(team.other(), toSquare)
    }

    fun moveIsOnShareRace(team: Team, fromSquare: Square, toSquare: Square): Boolean {
        return teamHasCounterOn(team, fromSquare)
                && empty(toSquare)
                && fromSquare.isRaceSquare()
    }

    fun empty(square: Square): Boolean {
        return !counters.containsKey(square)
    }

    fun teamHasCounterOn(team: Team, square: Square): Boolean {
        return counters[square] == team
    }

    fun completedCount(team: Team): Int {
        return completedCounters.getValue(team)
    }

    fun countersForTeam(team: Team): Set<Square> {
        return counters.entries
            .filter {e -> team === e.value }
            .map {e -> e.key}
            .toSet()
    }

    operator fun get(square: Square): Team? {
        return counters[square]
    }

    fun occupied(square: Square): Boolean {
        return counters.containsKey(square)
    }

    fun currentTeam(): Team {
        return currentTeam
    }

    fun swapTeam() {
        currentTeam = currentTeam.other()
    }

    fun allCountersStarted(team: Team): Boolean {
        return COUNTERS_PER_PLAYER == completedCounters[team]
    }

    fun unstartedCount(team: Team): Int {
        val inProgressCounters: Int = counters.entries.count { e -> team === e.value }
        val finishedCounters = completedCounters.getValue(team)
        return COUNTERS_PER_PLAYER - (finishedCounters + inProgressCounters)
    }

    fun allStartedOrComplete(team: Team): Boolean {
        return 0 == unstartedCount(team)
    }

    fun inPlayCount(team: Team): Int {
        return counters.values.count { t: Team -> team === t }
    }

    /** All counters currently on the board from both teams.  */
    fun inPlayCount(): Int {
        return counters.size
    }

    fun move(fromSquare: Square, newSquare: Square, team: Team) {
        counters.remove(fromSquare)
        if (newSquare !== Square.Off_board_finished) {
            counters[newSquare] = team
        } else {
            completedCounters[team] = 1 + completedCounters.getValue(team)
        }
    }

    fun allCompleted(team: Team): Boolean {
        return COUNTERS_PER_PLAYER == completedCounters[team]
    }

    companion object {
        const val COUNTERS_PER_PLAYER = 7
    }
}